Tweaks from the last few years

CDAGaming opened the pull request:
I run a minecraft mod called CraftPresence that, until recently, was using a jar-in-jar'd fork of this implementation of Discord's IPC system.

There will probably be quite a few questions, of which I'd be glad to answer, though I still plan to keep maintaining this fork so long as mod my mod keeps growing. Thank you again for making this amazing tool, as it's been fun to keep it going for awhile +1

Here is a list of changes that have come over time, in a greatly summarized form:
- Added `autoRegister` support (`WinRegistry` class added for Windows, tested up to Java 17/18)
- Updated JUnixSocket to 2.4.0, technically superseeding #20
  - Technically, this can go higher, but steps have been taken to retain support for Java 7, so we cannot go higher at least in this scope.
 - Use GSON instead of org.json
  - Same reason as prior note, game environments can more easily implement GSON (Or already have it included)
 - Added individualized logging for a debug or verbose mode that could be helpful in debugging and filtering out logs for user-level
 - Now uses `discord.com` instead of `discordapp.com` where applicable (New endpoint)
 - Added support for `partyPrivacy` and the `buttons` array
 - Added support for MacOS, as well as locating the `ipc pipe` from the Snap and Flatpak Distributions of Discord
   - Note that Flatpak'd JVMs are unable to locate the IPC Pipe, unless also using a Flatpak distribution of Discord, due to sandboxing restrictions.

Shengaero committed:
Fixes to pipeline closing

CDAGaming committed:
[Change] Update Dependencies

CDAGaming committed:
[Change] Remove SLF4J and Dependency Updates

CDAGaming committed:
Code Cleanup

CDAGaming committed:
[Change] Refactors

CDAGaming committed:
[Fix] Fixes

CDAGaming committed:
[Change] Added Debug Mode

CDAGaming committed:
[Change] Update Demo

CDAGaming committed:
[Fix] Fix IPC not looping on Windows

CDAGaming committed:
[Change] Convert to Google GSON

CDAGaming committed:
Merge remote-tracking branch 'remote/master'

- Also sync with CraftPresence's contents over the last two years

CDAGaming committed:
[Change] Optimized Constructors for IPCClient

CDAGaming committed:
[Change] More constructor optimization

CDAGaming committed:
[Change] Import and Docs cleanups from refactors

CDAGaming committed:
Suppress an IDE warning

CDAGaming committed:
[Fix] Fixed Java 7 Support

- Broke it during initial refactoring

CDAGaming committed:
[Fix] Fixed docs for Pipe#receive

CDAGaming committed:
[Revert] Go back to SLF4J

- I truly don't know what past me was on going away from this, chief.

CDAGaming committed:
[Change] Allow setting a different logger instance, if needed

CDAGaming committed:
Minor fixes

CDAGaming committed:
[Change] Migrate from Maven -> Gradle

CDAGaming committed:
[Fix] Do not include other JVM versions of classes in shadow jars

CDAGaming committed:
[Fix] Fixed Test Build unable to log by providing a sample binding

CDAGaming committed:
[Change] Comment out test bindings on default

- shadowJar likes to shadow everything, including runtimeOnly stuff

CDAGaming committed:
[Fix] Fixed shading

CDAGaming committed:
[Change] Remove Gradle publish logic

- Will use an alternative method for maven

CDAGaming committed:
[Fix] Fixed and sync some data

jagrosh commented:
There's a lot to unpack here. While there are plenty of changes in this that I'd love to see get merged in, there's also a lot that I don't think should be added. This isn't a comprehensive list or review, but just a start for some of the things I think should be removed before this could be merged:
- Git/env/ci-related files, which seem tied more to your environment or build process than this project.
- Gradle, I find for the scope of this project maven is simpler to use.
- Main.java, which shouldn't be present in a library. JUnit tests would be fine if you need an executable test file.
- [ ... more things as I notice them ... ]

CDAGaming commented:
I'm definitly happy to take feedback for this, though for what you've given so far:
- For the Main.java, I can easily remove that since the README already takes care of example usage
- The Gradle and CI/Env stuff I can easily split off as well (As I have a separate `publish` branch where that can go into)

Either way, definitely let me know of any more feedback. It definitely is a lot to sift through 😄 
I'll start making changes hopefully later today/tomorrow to go along with your review, mostly since I know there's likely to be more feedback.

CDAGaming committed:
Revert "[Change] Migrate from Maven -> Gradle"

This reverts commit c814f857b541f5c69c826547e40883245c8db413.

CDAGaming committed:
Revert test class addition

CDAGaming committed:
Merge remote-tracking branch 'remote/master'

CDAGaming committed:
[Fix] Fixed exceptions with `ACTIVITY_JOIN_REQUEST` + null user icon

CDAGaming committed:
[Fix] Fixed incorrect `cmd` for Rejecting/Ignoring Join Requests

CDAGaming committed:
[Change] Updates for Pomelo API Changes

Ref: https://discord.com/developers/docs/change-log#unique-usernames-on-discord

- Added null-safety for the discriminator field (Now defaults to `0` if no longer present in the Json)
- Fixed `getDefaultAvatarId` calculation to support non-discriminator lookups

- `global_name` is not being supported at this time, due to its infancy in the API. It will be added in a future commit once it becomes present in the API.

CDAGaming commented:
@jagrosh Any way we can get the ball rolling again on this PR, considering the recent fixes being made and the breaking API change discord has planned with the Pomelo Updates.

CDAGaming committed:
[Change] Merge changes from `plugin-dev`

- Allow freely changing the Debug and Verbose Log Settings (Primarily for better debugging mid-app-run / frontend configurability)
- Make IPCClient-Reader a daemon thread to avoid halting JVM exit

CDAGaming committed:
[Change] Implement re-connecting to IPC Pipe if disconnected

- This change integrates the Backoff logic from the Official RPC SDK, so that in the event of a disconnection, the Client will attempt to reconnect in subsequent `connect` executions, with increasing delays between `0.5 seconds` and `one minute`
- This only changes the behavior for an unexpected disconnection (Such as sending a Presence update while not connected to the RPC), it does not impact  a manual close (`onClose`)

CDAGaming committed:
[Change] Added support for `User#global_name`

- Added a `nickname` field to the user object, with a getter
- `name` field renamed to `username` (`getName` remains the same name for compatibility)
- Added `getEffectiveName` to retrieve the best display name (`username` if nickname is null, otherwise `nickname`)

CDAGaming committed:
[Fix] Fixed a cast exception that can occur with `discriminator` or `global_name`

- This further refines the future-proofing, and fixes an issue that can occur with no global name

CDAGaming committed:
[Change] Updated Dependencies + Fix `WinRegistry` support on JVM 11+

- Use Lenni Reflect to reflect without needing add-opens
- Also updated plugins, and recall Java 7 Support

CDAGaming committed:
[Fix] Bump Lenni Reflect to resolve some parity issues

CDAGaming committed:
[Change] Updated Dependencies

- Updated SLF4J to 2.0.10 (Was 2.0.7)
- Updated JUnixSocket to 2.8.3 (Was 2.7.0)
- Updated Lenni Reflect to 1.3.2 (Was 1.2.4)
- Updated Maven Javadoc Plugin to 3.6.3 (Was 3.5.0)

CDAGaming committed:
[Change] Updated Dependencies

- Updated SLF4J-API to 2.0.12 (Was 2.0.10)

CDAGaming committed:
[Change] Updated Dependencies

- Updated JUnixSocket to 2.9.0 (Was 2.8.3)

CDAGaming committed:
[Change] Updated Dependencies

- Updated JUnixSocket to 2.9.1 (Was 2.9.0)

CDAGaming committed:
[Change] Updated Dependencies

- Updated SLF4J to 2.0.13 (Was 2.0.12)

CDAGaming committed:
[Change] Updated Dependencies

- Updated Lenni Reflect (1.3.2 -> 1.3.4)

CDAGaming committed:
[Change] Updated Dependencies

- Updated GSON (2.10.1 -> 2.11.0)

CDAGaming committed:
[Fix] Ensure slashes are escaped

- Should fix support for Java 22 on Windows, where this issue came up

CDAGaming committed:
[Change] Remove unnecessary `path` null check

CDAGaming committed:
[Fix] Minor cleanups for `Exception | Error` catch calls

CDAGaming committed:
[Fix] Fixed a javadoc exception with `IPCClient#connect`

- `Thread.sleep` call now matches all the others

CDAGaming committed:
[Change] Adjusted some error messages

- Adjusted the `WindowsPipe` to throw a proper `RuntimeException` in cases where we are unable to write to a IPC pipe (There is a difference from `exists` to `can read/write`)
  - A commit awhile ago broke the prior behavior, allowing the Pipe to be written to, causing another exception later. This change lets it fail more safely and as expected
- Clarified the error message for when an `IOException` is thrown for the Mac/Unix pipes
  - Its a bit vague since it triggers under almost any type of operation failure
- Replaced a `printStackTrace` call in `IPCClient#connect` for autoRegister behavior, so it'll now use the logger for the `debugMode` version
- Clarified error message for if the `WinRegistry` constructor fails

CDAGaming committed:
[Change] Removed `ExtendedLong`

- No longer necessary

CDAGaming committed:
[Change] `IPCClient#sendRichPresence` cannot take in a null `presence`

CDAGaming committed:
[Change] Added an `equals` and `hashCode` method for `RichPresence` obj

CDAGaming committed:
[Change] Removed `DataConsumer`

- No longer necessary

CDAGaming committed:
[Change] Code Cleanup

- Corrected documentation for several files
- Enforce Java 8+ behaviors, fully dropping Java 7 support (As tools exist now to handle downgrading)

CDAGaming committed:
[Fix] Fixed a typo

CDAGaming committed:
[Fix] Fixed a missing default avatar icon for the `pink` icon

- Added the 6th Default Avatar ID, which is only available to Migrated users
- Fixed `getDefaultAvatarId` for legacy users, and to more clearly represent discord documentation

CDAGaming committed:
[Fix] Fix discrepancy in `autoRegister` functionality

- Fixed incorrect `commandKeyName` allocation

CDAGaming committed:
[Change] Updated Dependencies

- Updated JUnixSocket (`2.9.1` -> `2.10.0`)
- Updated maven-javadoc-plugin (`3.6.3` -> `3.8.0`)

CDAGaming committed:
[Change] Added data sanity for `Party Privacy`

- Ported from another project I created

CDAGaming committed:
[Change] Added support for `ActivityType`

- Ref: https://github.com/discord/discord-api-docs/pull/7033

CDAGaming committed:
[Change] Updated Dependencies

- Updated SLF4J to 2.0.16
- Updated JUnixSocket to 2.10.1
- Updated Maven Javadoc Plugin to 3.10.1

CDAGaming committed:
[Change] Updated Dependencies

- Updated LenniReflect to 1.4.0
- Updated maven-javadoc-plugin to 3.11.1

CDAGaming committed:
[Change] Updated Dependencies

- Updated Gson to 2.13.1
- Updated SLF4J to 2.0.17

CDAGaming committed:
[Change] Updated Dependencies

- Updated LenniReflect to 1.5.0

CDAGaming committed:
[Change] Added support for several new RPC fields

  - `status_display_type`: Allows determining whether to display the `Name`, `State`, or `Details` section of the RPC as the short activity text
    - Example: `<activity_type> <status_display_type_message>` => `Playing Minecraft`
  - Added URL support for the `state`, `details`, `largeImage`, and `smallImage` RPC fields
    - BREAKING: `set<Large,Small>Image(imageKey, imageText)` => `set<Large,Small>ImageWithTooltip(imageKey, imageText)`
  - `name`: Allows setting a custom name for your application, defaults to the one specified in Discord if left empty

CDAGaming committed:
[Change] Updated Dependencies

- Updated LenniReflect to 1.6.1
- Updated Maven Source Plugin to 3.4.0

CDAGaming committed:
[Change] Re-add null support for `sendRichPresence` calls

- This turns out to be valid API usage, and due to recent discord breakage, is now required when closing the IPC

CDAGaming committed:
[Fix] Improved null-safety for Pipe handshake data and RPC enum defaults

- Defined proper enum defaults for the RPC Builder
- Throw proper exception if the IPC client was valid, but threw null data

Partial merge of https://github.com/CDAGaming/DiscordIPC/pull/2
